bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -2222,9 +2223,22 @@ private boolean updateMetadataIfPossible(LedgerMetadata metadata, LedgerMetadata
             metadata.setVersion(newMeta.getVersion());
             // merge ensemble infos from new meta except last ensemble
             // since they might be modified by recovery tool.
+
             metadata.mergeEnsembles(newMeta.getEnsembles());
-            writeLedgerConfig(new ChangeEnsembleCb(ensembleInfo, curBlockAddCompletions,
-                    ensembleChangeIdx));
+            if (!areFailedBookiesReplaced(metadata, ensembleInfo)) {
+                // If the in-memory data contains the failed bookie, someone else
+                // merged the metadata and reinstated old copy. Let's attempt to
+                // replace the bookie again.
+                try {
+                    ensembleInfo = replaceBookieInMetadata(ensembleInfo.failedBookies, numEnsembleChanges.get());
+                } catch (BKException.BKNotEnoughBookiesException e) {
+                    LOG.error("Could not get additional bookie to remake ensemble, closing ledger: {}", ledgerId);
+                    handleUnrecoverableErrorDuringAdd(e.getCode());
```

Comment:
I think if we return false here ReReadLedgerMetadataCb will already handle calling handleUnrecoverableErrorDuringAdd, so doing it here results in it being called twice.

Reply:
true; I will remove the handleUnrecoverableErrorDuringAdd() from here. There are other places in this method returns false but won't call handleUnrecoverableErrorDuringAdd.